### PR TITLE
New detection models

### PR DESCRIPTION
**Pull request by basedrhys**
This adds classes for using Detectron2 models, as well as TF models directly with tensorflow (rather than OpenCV). Note that to use these models you will need to install the frameworks as per their install guide.

### CODE REVIEW

**Comment by nicholaskajoh** [comment]
Hi @basedrhys! I'll review and share my feedback later today. Cheers!

File: detectors/detector.py [hunk 1]
```diff
@@ -12,6 +12,8 @@ def get_bounding_boxes(frame, model):
         from detectors.haarcascade import get_bounding_boxes as gbb
     elif model == 'tfoda':
         from detectors.tfoda import get_bounding_boxes as gbb
+    elif model == 'detectron2':
```
Comment (nicholaskajoh) [inline detectors/detector.py]: Can you provide a sample model with which to test this detector?

File: detectors/detectron2.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
+'''
+Performs vehicle detection using models created with the YOLO (You Only Look Once) neural net.
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: I believe this comment should describe Detectron2 and not YOLO.

File: detectors/detectron2.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import ast
+import os
+import cv2
+import torch
+
+import detectron2
+from detectron2.utils.logger import setup_logger
+setup_logger()
+
+# import some common libraries
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: Uncessary comment

File: detectors/detectron2.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+import numpy as np
+import random
+
+# import some common detectron2 utilities
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: This comment isn't necessary too

File: detectors/detectron2.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+https://pjreddie.com/darknet/yolo/
+'''
+
+import ast
+import os
+import cv2
+import torch
+
+import detectron2
+from detectron2.utils.logger import setup_logger
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: It would be nice to group detectron2 imports together

File: detectors/detectron2.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import numpy as np
+import random
+
+# import some common detectron2 utilities
+from detectron2.engine import DefaultPredictor
+from detectron2.config import get_cfg
+from detectron2.utils.visualizer import Visualizer
+from detectron2.data import MetadataCatalog
+
+with open(os.getenv('DETECTRON2_CLASSES_PATH'), 'r') as classes_file:
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: Please add these environment variables to .env.example

File: detectors/detectron2.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    CLASSES = dict(enumerate([line.strip() for line in classes_file.readlines()]))
+with open(os.getenv('DETECTRON2_CLASSES_OF_INTEREST_PATH'), 'r') as coi_file:
+    CLASSES_OF_INTEREST = tuple([line.strip() for line in coi_file.readlines()])
+
+use_gpu = ast.literal_eval(os.getenv('ENABLE_GPU_ACCELERATION'))
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: This variable is not used anywhere in your code

File: detectors/detectron2.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+conf_threshold = float(os.getenv('DETECTRON2_CONFIDENCE_THRESHOLD'))
+
+# # initialize model with weights and config
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: This is nitpicking but I'd prefer if you removed the extra `#`

File: detectors/detectron2.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+cfg = get_cfg()
+cfg.merge_from_file(os.getenv('DETECTRON2_CONFIG_PATH'))
+cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.5  # set threshold for this model
+# Find a model from detectron2's model zoo. You can either use the https://dl.fbaipublicfiles.... url, or use the following shorthand
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: I don't understand what this comment is saying. Why is the url incomplete?

File: detectors/detectron2.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+cfg.MODEL.WEIGHTS = os.getenv('DETECTRON2_WEIGHTS_PATH')
+cfg.MODEL.ROI_HEADS.NUM_CLASSES = int(os.getenv('DETECTRON2_NUM_CLASSES'))  # only has one class (ballon)
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: This comment is uncessary. Remember this detector is going to be used with models trained on a variety of classes.

File: detectors/detectron2.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = float(os.getenv('DETECTRON2_CONFIDENCE_THRESHOLD'))
+cfg.DATALOADER.NUM_WORKERS = 2
+
+if not torch.cuda.is_available():
+    print("No GPU available, using CPU")
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: Use the logger function instead of print (level for this one should be debug).

File: detectors/detectron2.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+predictor = DefaultPredictor(cfg)
+
+def convert_box_to_array(this_box):
+    res = []
+    for val in this_box:
+        for ind_val in val:
+            res.append(float(ind_val))
+
+    # Convert x2, y2 into w, h
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: This says very little about what's going on. Either provide higher-level info or remove.

File: detectors/detectron2.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+if not torch.cuda.is_available():
+    print("No GPU available, using CPU")
+    cfg.MODEL.DEVICE = 'cpu'
+else:
+    print("GPU Available, using GPU")
+    cfg.MODEL.DEVICE = 'cuda'
+
+predictor = DefaultPredictor(cfg)
+
+def convert_box_to_array(this_box):
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: What does this function do? A doc block would be great!

File: detectors/detectron2.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+    for i, pred in enumerate(outputs["instances"].pred_classes):
+        coco_class = int(pred)
+        class_string = CLASSES[coco_class]
+        _classes.append(class_string)
+
+        score = round(float(outputs['instances'].scores[i]), 3)
+        _confidences.append(score)
+
+
```
Comment (nicholaskajoh) [inline detectors/detectron2.py]: Remove extra new line

File: detectors/detector.py [hunk 15]
```diff
@@ -14,6 +14,8 @@ def get_bounding_boxes(frame, model):
         from detectors.tfoda import get_bounding_boxes as gbb
     elif model == 'detectron2':
         from detectors.detectron2 import get_bounding_boxes as gbb
+    elif model == 'tfoda_new':
```
Comment (nicholaskajoh) [inline detectors/detector.py]: I'd actually like for this to replace the existing tfoda implementation. Perhaps, a separate PR would be better?

File: detectors/detector.py [hunk 16]
```diff
@@ -14,6 +14,8 @@ def get_bounding_boxes(frame, model):
...
+        from detectors.tfoda_new import get_bounding_boxes as gbb
     else:
         raise Exception('Invalid detector model, algorithm or API specified (options: yolo, tfoda, haarcascade)')
```
Comment (nicholaskajoh) [inline detectors/detector.py]: Update error message appropriately

File: detectors/tfoda_new.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+    res.append((box[3] - box[1]) * img_w) #width
+    res.append((box[2] - box[0]) * img_h) #height
+
+    return res
+
+with open(os.getenv('TFODA_CLASSES_PATH'), 'r') as classes_file:
+    CLASSES = dict(enumerate([line.strip() for line in classes_file.readlines()]))
+
+confidence_threshold = float(os.getenv('TFODA_CONFIDENCE_THRESHOLD'))
+model_dir = os.path.join(os.getenv('TFODA_WEIGHTS_PATH'), "saved_model")
```
Comment (nicholaskajoh) [inline detectors/tfoda_new.py]: Shouldn't `saved_model` be in `TFODA_WEIGHTS_PATH`? Remember that anyone with any directory structure should be able to use this without having to change the code.

File: detectors/tfoda_new.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+
+model = tf.saved_model.load(str(model_dir))
+model = model.signatures['serving_default']
+
+def get_bounding_boxes(image):
+    image = image[:, :, ::-1]
+    # print(type(img))
+    # the array based representation of the image will be used later in order to prepare the
+    # result image with boxes and labels on it.
+    # img_w, img_h = img.size
```
Comment (nicholaskajoh) [inline detectors/tfoda_new.py]: Please remove ALL commented code. If the code is not going to run, it shouldn't be here.

**Comment by nicholaskajoh** [comment]
Thanks for the time and effort you put into this @basedrhys. In my mind, this should be 3 PRs because the 3 commits do 3 different things.

PS: If you're up for it, there's a backlog of interesting work to be done on Ivy including model training, vehicle color detection, vehicle speed estimation, ANPR etc.

**Comment by basedrhys** [comment]
Ok, I'll make those changes and split it up into separate PRs.

I'd love to however my application of this is around simply counting animals, so I'm not in desperate need for those additions.

Cheers

File: detectors/detector.py (hunk 1 above)
Comment (luiscosio) [reply detectors/detector.py]: Cityscapes or Pascal VOC seems to give good results: https://github.com/facebookresearch/detectron2/blob/master/MODEL_ZOO.md#cityscapes--pascal-voc-baselines

File: detectors/detectron2.py (hunk 6 above)
Comment (luiscosio) [reply detectors/detectron2.py]: This are the required .env variables:

# DETECTRON2
DETECTRON2_CONFIDENCE_THRESHOLD=0.8
DETECTRON2_CONFIG_PATH="./data/detectors/detectron2/model_final_af9cf5.cfg"
DETECTRON2_WEIGHTS_PATH="./data/detectors/detectron2/model_final_af9cf5.pkl"
DETECTRON2_NUM_CLASSES=8
DETECTRON2_CLASSES_PATH="./data/detectors/detectron2/classes.txt"
DETECTRON2_CLASSES_OF_INTEREST_PATH="./data/detectors/detectron2/classes_of_interest.txt"

**Comment by luiscosio** [comment]
Thank you @basedrhys, works like a charm!

File: detectors/detector.py (hunk 1 above)
Comment (nicholaskajoh) [reply detectors/detector.py]: Thanks @luiscosio

**Comment by basedrhys** [comment]
Hi, 
I've made those changes so the PR should be updated with them.
Do you still want them as seperate PRs? It seems like unnecessary extra work for both of us (please correct me if I'm wrong, I'm a noob with PRs).

**Comment by nicholaskajoh** [comment]
Hey @basedrhys,

I prefer each feature/fix to be in a separate PR because it makes it easier to keep track of changes to the codebase. However I understand it's difficult to separate things right now so I'll permit this. But for next time in any project you might be contributing to, ensure a PR does only one unit of work.

I'll review the changes later today and get back to you.

Well done!

**Comment by basedrhys** [comment]
Hi @nicholaskajoh,

Yeah that makes sense! I'll be a bit more careful in the future.

Cheers

**Comment by nicholaskajoh** [comment]
This looks good! I'll test locally, make a few tweaks and merge over the weekend.

**Comment by nicholaskajoh** [comment]
Hi @basedrhys!

Sorry "over the weekend" turned out to be three weeks. I didn't have the time to test and tweak the code until now.

I've merged! Did so locally (https://github.com/nicholaskajoh/ivy/commit/2709c64d672f0ecb8c7d2453c24ed4c11f45a317) so I'll be closing this.

PS: You did a great job. Thanks!

Warm regards.

**Pull request closed without merging**